# HG changeset patch
# User Pantxo Diribarne <pantxo.diribarne@gmail.com>
# Date 1546850417 -3600
#      Mon Jan 07 09:40:17 2019 +0100
# Branch stable
# Node ID 6f22d0c91beb674ea926fb19fbb2c96d508027de
# Parent  310143b2c2c437f3397ddd68e75d116504fa68bb
Fix memory leak with FT_Glyph objects (bug #55287)

* ft-text-renderer.cc (ft_text_renderer::process_character): Call FT_Done_Glyph
  after we are done with a loaded glyph.
  (ft_text_renderer::ft_text_renderer): Initialize line_xoffset as per PVS
  analyzer notice.

diff --git a/libinterp/corefcn/ft-text-renderer.cc b/libinterp/corefcn/ft-text-renderer.cc
--- a/libinterp/corefcn/ft-text-renderer.cc
+++ b/libinterp/corefcn/ft-text-renderer.cc
@@ -371,18 +371,18 @@ namespace octave
       ROTATION_270 = 3
     };
 
   public:
 
     ft_text_renderer (void)
       : base_text_renderer (), font (), bbox (1, 4, 0.0), halign (0),
         xoffset (0), line_yoffset (0), yoffset (0), mode (MODE_BBOX),
-        color (dim_vector (1, 3), 0), m_ymin (0), m_ymax (0), m_deltax (0),
-        m_max_fontsize (0)
+        color (dim_vector (1, 3), 0), line_xoffset (0), m_ymin (0), m_ymax (0),
+        m_deltax (0), m_max_fontsize (0)
     { }
 
     // No copying!
 
     ft_text_renderer (const ft_text_renderer&) = delete;
 
     ft_text_renderer& operator = (const ft_text_renderer&) = delete;
 
@@ -835,16 +835,17 @@ namespace octave
                     FT_BBox  glyph_bbox;
                     FT_Glyph_Get_CBox (glyph, FT_GLYPH_BBOX_UNSCALED,
                                        &glyph_bbox);
                     m_deltax = (glyph_bbox.xMax - face->glyph->advance.x) >> 6;
                     m_ymin = math::min ((glyph_bbox.yMin >> 6) + yoffset,
                                         m_ymin);
                     m_ymax = math::max ((glyph_bbox.yMax >> 6) + yoffset,
                                         m_ymax);
+                    FT_Done_Glyph (glyph);
                     update_line_bbox ();
                   }
                 break;
               }
           }
       }
 
     return glyph_index;
